src/main/java/com/google/sps/recommendation/ProcessNearestNeighbors.java

```diff
@@ -0,0 +1,125 @@
+package com.google.sps.recommendation;
+
+import org.deeplearning4j.models.embeddings.inmemory.InMemoryLookupTable;
+import org.deeplearning4j.models.embeddings.wordvectors.WordVectors;
+import org.deeplearning4j.models.embeddings.loader.WordVectorSerializer;
+import org.deeplearning4j.models.word2vec.wordstore.VocabCache;
+
+import org.nd4j.linalg.api.buffer.DataType;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.factory.Nd4j;
+import org.nd4j.linalg.indexing.NDArrayIndex;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.stream.IntStream;
+import com.google.sps.tool.ArrayUtil;
+import org.mapdb.*;
+
+/* This function of this file is to ONLY pre-process the nearest neigbors of 300,000 words.
+*  It will create/update a binary .db file. 
+*  
+*  Processing Time: 
+*  This took about 12 hours to process the nearest neighbors. 
+*  Performance analysis and stress testing were done while developing this algorithm.
+*  It turns out that computing the nearest neighbors for a batch size of 500 words was 
+*  approximately optimal in order to avoid memory failure. It's important to keep in mind that 
+*  using the N4dj API matrix operations result in slow performances
+*  compared to developing a similar program in Java with numpy. 
+*  These time results and optimal batch size are based on running this file on a Google Pixelbook. 
+*  It may change for different hardwares.
+*
+*  Time example: 
+*  Each batch of 500 words takes about 60,000 - 80,000 ms. 
+*  Processing 100 batches can take from 1 hour 30 minutes - 2 hours. 
+*  Some underlying reasons for this slow time performance:
+*  1) N4dj's behind the hood operations. For example, to slice a large matrix by a range of indices,
+*     N4dj has to make a copy which is really slow, especially with a big matrix
+*  2) Each computation of a batch results in sorting the 
+*     cosine similarity values in each row of size 300,000
+*  3) There is some latency with serializing the result into the .db file. 
+*     Note: We use MapDB API to serialize a HashMap Object, which utilizes Random Access Files (RAF)
+*/
+public class ProcessNearestNeighbors {
+  public static void main(String[] args) throws IOException {
+    WordVectors vec =
+        WordVectorSerializer.readWord2VecModel(
+            "/home/ngothomas/GoogleNews-vectors-negative300-SLIM.bin.gz");
+    InMemoryLookupTable lookupTable = (InMemoryLookupTable) vec.lookupTable();
+    VocabCache lookupCache = (VocabCache) vec.vocab();
+
+    /* Create HashMap that maps word to index,
+    *  where the index represents where the word is
+    *  in the word vector matrix 300,000 x 300 */
+    Map<String, Integer> wordToIndex = new HashMap<>();
+    Map<Integer, String> indexToWord = new HashMap<>();
```

Comment:
Maps from integers to strings could just be arrays, especially if the integers are dense.

Reply:
true, this would probably be faster.